agent/lib/kontena/rpc/docker_container_api.rb

```diff
@@ -109,18 +109,72 @@ def top(id)
       def exec(id, cmd, opts = {})
         container = Docker::Container.get(id)
         container.exec(cmd, opts)
-      rescue Docker::Error::NotFoundError => exc
+      rescue Docker::Error::NotFoundError
         raise RpcServer::Error.new(404, 'Not found')
       end
 
+      def create_exec(id)
+        container = Docker::Container.get(id)
+        executor = Kontena::Actors::ContainerExec.new(container)
+        actor_id = "container_exec_#{executor.uuid}"
+        Celluloid::Actor[actor_id] = executor
+        { id: executor.uuid }
+      rescue Docker::Error::NotFoundError
+        raise RpcServer::Error.new(404, 'Container not found')
+      end
+
+      def run_exec(id, cmd)
+        actor_id = "container_exec_#{id}"
+        executor = Celluloid::Actor[actor_id]
+        if executor
+          executor.async.run(cmd)
+        else
+          raise RpcServer::Error.new(404, "Exec session (#{id}) not found")
+        end
+
+        {}
+      end
+
+      def tty_exec(id, cmd)
+        actor_id = "container_exec_#{id}"
+        executor = Celluloid::Actor[actor_id]
+        if executor
+          executor.async.interactive(cmd)
+        else
+          raise RpcServer::Error.new(404, "Exec session (#{id}) not found")
+        end
+
+        {}
+      end
+
+      def tty_input(id, input)
+        actor_id = "container_exec_#{id}"
+        executor = Celluloid::Actor[actor_id]
+        if executor
+          executor.async.input(input)
+        end
+      end
+
+      def terminate_exec(id)
+        actor_id = "container_exec_#{id}"
+        executor = Celluloid::Actor[actor_id]
+        if executor
+          Celluloid::Actor.kill(executor) if executor.alive?
```

Comment:
Yes, it looks like the Docker API doesn't have anything to kill an exec, which is surprising.

If you SIGINT a `docker exec -i ... sleep 300` via Ctrl-c, then it just kills the CLI process, and leaves the exec process (`sleep 300`) running inside the container.

With `docker exec -it` with raw tty mode, a ^C results in "\x03" on the TTY, which the CLI passes through to the API, and gets delivered to the pseudotty within the container, signaling the process.